File: Collections/BinaryHeap.cs
using System;

namespace Collections
{
    /// <summary>
    /// Min-oriented Priority Queue using a Binary Heap with dynamically-resizing backing array.
    /// </summary>
    public class BinaryHeap<K> : IPriorityQueue<K> where K : class, IComparable<K>
    {
        private K[] heap;
        private int n = 0;  // The number of keys on the heap

        public BinaryHeap(int capacity = 10)
        {
            // A (minimum) heap-ordered complete binary tree is a binary tree which is completely balanced, 
            // except possibly for the lowest levels, and also one  where the parent node's key is
            // at most as large as that of its children. In the array representation:
            // - Parent of k is given by k/2.
            // - Children of k are 2*k, 2*k+1
            // - The heap invariant states heap[k] <= Min(heap[2*k], heap[2*k+1])
            if (capacity < 1)
            {
                throw new ArgumentOutOfRangeException();
            }

            heap = new K[capacity + 1];
        }

        public bool IsHeapOrdered()
        {
            // There are plenty of ways to write this, but the code from Algs. I 
            // refects the symmetry of the recursive solution
            bool IsHeapOrdered(int k)
            {
                if (k > n)
                    return true;
                var left = 2 * k;
                var right = 2 * k + 1;
                if (left <= n && Greater(k, left))
                    return false;
                if (right <= n && Greater(k, right))
                    return false;
                return IsHeapOrdered(left) && IsHeapOrdered(right);
            }
            return IsHeapOrdered(1);
        }

        public bool IsEmpty() => n == 0;

        public void Push(K key)
        {
            // Lazily increase the size of the backing array if necessary
            if (n == heap.Length - 1)
            {
                Resize(2 * heap.Length);
            }
            heap[++n] = key;
            Swim(n);
        }

        public K Peek()
        {
            if (IsEmpty())
            {
                throw new InvalidOperationException("Empty heap");
            }
            return heap[1];
        }

        public K Pop()
        {
            if (IsEmpty())
            {
                throw new InvalidOperationException("Empty heap");
            }
            // Lazily reduce the size of the backing array if necessary
            if (n == heap.Length / 4)
            {
                Resize(heap.Length / 2);
            }

            var smallest = heap[1];
            Swap(1, n);
            heap[n--] = null;  // No loitering
            Sink(1);
            return smallest;
        }

        public int Size() => n;

        public int Capacity() => heap.Length;

        private void Resize(int capacity)
        {
            var temp = new K[capacity];
            for (int i = 1; i <= n; i++)
            {
                temp[i] = heap[i];
            }
            heap = temp;
        }

        // Utilities to restore the heap invariant

        private void Sink(int k)
        {
            while (2 * k <= n)
            {
                int samllest = 2 * k < n && Greater(2 * k, 2 * k + 1) ? 2 * k + 1 : 2 * k;
                if (!Greater(k, samllest)) break;
                Swap(k, samllest);
                k = samllest;
            }
        }

        private void Swim(int k)
        {
            while (k > 1 && Greater(k / 2, k))
            {
                Swap(k / 2, k);
                k = k / 2;
            }
        }

        private bool Greater(int i, int j) => heap[i].CompareTo(heap[j]) > 0;

        private void Swap(int i, int j)
        {
            (heap[i], heap[j]) = (heap[j], heap[i]);
        }
    }

    public class BinaryHeap<K, V> : IPriorityQueue<K, V> where K : class, IComparable<K> where V : class
    {
        private K[] keys;
        private V[] values;
        private HashMap<V, int> index;
        private int n = 0;  // The number of keys on the heap

        public BinaryHeap(int capacity = 10)
        {
            if (capacity < 1)
            {
                throw new ArgumentOutOfRangeException();
            }

            keys = new K[capacity + 1];
            values = new V[capacity + 1];
            index = new HashMap<V, int>();
        }

        public bool IsHeapOrdered()
        {
            bool IsHeapOrdered(int k)
            {
                if (k > n) return true;
                var left = 2 * k;
                var right = 2 * k + 1;
                if (left <= n && Greater(k, left)) return false;
                if (right <= n && Greater(k, right)) return false;
                return IsHeapOrdered(left) && IsHeapOrdered(right);
            }
            return IsHeapOrdered(1);
        }

        public bool IsEmpty() => n == 0;

        public bool Contains(V value) => index.Contains(value);

        public void Push(V value, K key)
        {
            if (index.Contains(value))
            {
                throw new InvalidOperationException();  // Values must be unique
            }

            // Lazily increase the size of the backing array if necessary
            if (n == keys.Length - 1)
            {
                Resize(2 * keys.Length);
            }
            n++;
            keys[n] = key;
            values[n] = value;
            index[value] = n;  // Initially give this pair an the heapsize as an index
            Swim(n);
        }

        public void Update(V value, K key)
        {
            var i = index[value];
            var prev = keys[i];
            keys[i] = key;

            if (Greater(key, prev))
            {
                Sink(i);
            }
            else
            {
                Swim(i);
            }
        }

        public (K key, V value) Peek()
        {
            if (IsEmpty())
            {
                throw new InvalidOperationException("Empty heap");
            }
            return (keys[1], values[1]);
        }

        public (K key, V value) Pop()
        {
            if (IsEmpty())
            {
                throw new InvalidOperationException("Empty heap");
            }
            // Lazily reduce the size of the backing array if necessary
            if (n == keys.Length / 4)
            {
                Resize(keys.Length / 2);
            }

            var (key, value) = (keys[1], values[1]);
            Swap(1, n);
            n--;
            (keys[n], values[n]) = (null, null);  // No loitering
            Sink(1);
            index.Delete(value);  // Clear index

            return (key, value);
        }

        public int Size() => n;

        public int Capacity() => keys.Length;

        private void Resize(int capacity)
        {
            var newKeys = new K[capacity];
            var newValues = new V[capacity];
            for (int i = 1; i <= n; i++)
            {
                newKeys[i] = keys[i];
                newValues[i] = values[i];
            }
            keys = newKeys;
            values = newValues;
        }

        // Utilities to restore the heap invariant

        private void Sink(int k)
        {
            while (2 * k <= n)
            {
                int samllest = 2 * k < n && Greater(2 * k, 2 * k + 1) ? 2 * k + 1 : 2 * k;
                if (!Greater(k, samllest)) break;
                Swap(k, samllest);
                k = samllest;
            }
        }

        private void Swim(int k)
        {
            while (k > 1 && Greater(k / 2, k))
            {
                Swap(k / 2, k);
                k = k / 2;
            }
        }

        private bool Greater(int i, int j) => keys[i].CompareTo(keys[j]) > 0;

        private bool Greater(K k, K l) => k.CompareTo(l) > 0;

        private void Swap(int i, int j)
        {
            (keys[i], keys[j]) = (keys[j], keys[i]);
            (index[values[i]], index[values[j]]) = (j, i);  // Maintain the index during a swap
            (values[i], values[j]) = (values[j], values[i]);
        }
    }
}

File: src/main/java/in/bfdes/collections/BinaryHeap.java
package in.bfdes.collections;

import java.util.function.Predicate;

public class BinaryHeap<V, K extends Comparable<K>> implements PriorityQueue<V, K> {
    private K[] keys;
    private V[] values;
    private final Map<V, Integer> index = new HashMap<>();
    private int size = 0;

    public BinaryHeap() {
        this(10);
    }

    public BinaryHeap(int initialCapacity) {
        if (initialCapacity < 1)
            throw new IllegalArgumentException();
        keys = (K[]) new Object[initialCapacity + 1];
        values = (V[]) new Object[initialCapacity + 1];
    }

    @Override
    public boolean contains(V value) {
        return index.contains(value);
    }

    @Override
    public void push(V value, K key) {
        if (contains(value)) {
            var i = index.get(value);
            var oldKey = keys[i];
            keys[i] = key;
            var isGreater = key.compareTo(oldKey) > 0;
            if (isGreater)
                sink(i);
            else
                swim(i);
        } else {
            if (isFull())
                resize(2 * keys.length);
            keys[++size] = key;
            values[size] = value;
            index.put(value, size);
            swim(size);
        }
    }

    @Override
    public V peek() {
        if (isEmpty())
            throw new IllegalStateException();
        return values[1];
    }

    @Override
    public V pop() {
        if (isEmpty())
            throw new IllegalStateException();
        var value = values[1];

        swap(1, size);
        keys[size] = null;
        values[size] = null;
        size--;
        sink(1);
        index.delete(value);

        if (size == keys.length / 4)
            resize(keys.length / 2);

        return value;
    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    private boolean isFull() {
        return size == keys.length - 1;
    }

    /**
     * Returns `true` if this binary tree is heap ordered, `false` otherwise.
     * Note: A `BinaryHeap` should always be in heap order -- this method is for debugging.
     */
    public boolean isHeapOrdered() {
        var isHeapOrdered = new Predicate<Integer>() {
            @Override
            public boolean test(Integer k) {
                if (k > size)
                    return true;
                var l = 2 * k;
                var r = 2 * k + 1;
                if (l <= size && isGreater(k, l))
                    return false;
                if (r <= size && isGreater(k, r))
                    return false;
                return test(l) && test(r);
            }
        };
        return isHeapOrdered.test(1);
    }

    private void resize(int newCapacity) {
        var newKeys = (K[]) new Object[newCapacity];
        var newValues = (V[]) new Object[newCapacity];
        for (var i = 0; i < size; i++) {
            newKeys[i] = keys[i];
            newValues[i] = values[i];
        }
        keys = newKeys;
        values = newValues;
    }

    private void sink(int i) {
        while (2 * i <= size) {
            var l = 2 * i;
            var r = 2 * i + 1;

            var m = l < size && isGreater(l, r) ? r : l;
            if (!isGreater(i, m))
                break;
            swap(i, m);
            i = m;
        }
    }

    private void swim(int i) {
        while (i > 1 && isGreater(i / 2, i)) {
            swap(i / 2, i);
            i = i / 2;
        }
    }

    private boolean isGreater(int i, int j) {
        return keys[i].compareTo(keys[j]) > 0;
    }

    private void swap(int i, int j) {
        var tmpKey = keys[i];
        keys[i] = keys[j];
        keys[j] = tmpKey;

        index.put(values[i], j);
        index.put(values[j], i);

        var tmpValue = values[i];
        values[i] = values[j];
        values[j] = tmpValue;
    }
}
